django_project/realtime/tasks/earthquake.py

```diff
@@ -104,8 +167,108 @@ def generate_event_report(earthquake_event):
     :type earthquake_event: Earthquake
     :return:
     """
-    if earthquake_event.hazard_layer_exists:
-        pass
-        # TODO: Generate Earthquake report
+    if not earthquake_event.hazard_layer_exists:
+
+        # Skip all process if exists
+        return
+
+    # Check raw hazard stored
+    elif not earthquake_event.shake_grid_xml:
+
+        earthquake_event.refresh_from_db()
+
+        # Store Grid XML from hazard path into db
+        if earthquake_event.shake_grid:
+            shake_grid_xml = earthquake_event.shake_grid.read()
+
+        # Attempt to get grid.xml by guessing from hazard path
+        else:
+            dir_name = os.path.dirname(earthquake_event.hazard_path)
+            grid_path = os.path.join(dir_name, GRID_FILE_DEFAULT_NAME)
+            with open(grid_path) as f:
+                shake_grid_xml = f.read()
+
+        # Use save to trigger signals again
+        Earthquake.objects.filter(
+            id=earthquake_event.id).update(
+            shake_grid_xml=shake_grid_xml)
+
+        earthquake_event.refresh_from_db()
+
+        # Remove un-needed Grid XML
+        if earthquake_event.shake_grid:
+            earthquake_event.shake_grid.delete(save=False)
+
+        earthquake_event.save()
+
+    # Check impact layer
+    elif (not earthquake_event.impact_layer_exists and
+            earthquake_event.need_run_analysis):
+
+        # If hazard exists but impact layer is not, then create a new analysis
+        # job.
+        run_earthquake_analysis(earthquake_event)
+
+    # Check report
+    elif (not earthquake_event.has_reports and
+            earthquake_event.need_generate_reports):
+
+        # If analysis is done but report doesn't exists, then create the
+        # reports.
+        generate_earthquake_report(earthquake_event)
+
+
+def run_earthquake_analysis(event):
+    """Run analysis.
+
+    :param event: event instance
+    :type event: Earthquake
+    """
+    event.refresh_from_db()
+    hazard_layer_uri = event.hazard_path
+    async_result = run_multi_exposure_analysis.delay(
+        hazard_layer_uri,
+        EARTHQUAKE_EXPOSURES,
+        EARTHQUAKE_AGGREGATION,
+    )
+    Earthquake.objects.filter(id=event.id).update(
+        analysis_task_id=async_result.task_id,
+        analysis_task_status=async_result.state)
+
+
+def generate_earthquake_report(event):
+    """Generate report for event.
+
+    :param event: event instance
+    :type event: Earthquake
+    """
+    event.refresh_from_db()
+    impact_layer_uri = event.impact_file_path
+
+    result = AsyncResult(id=event.analysis_task_id, app=headless_app)
+
+    layer_order = []
```

Comment:
Check this out, @ismailsunni , I'm trying to dynamically substitute layer order with layers in the multiexposure analysis results. Do you have any input on this?

Reply:
Nice. It will work for layer from output. But for hazard layer (flood data and volcanic ash hazard layer) I will use the current approach.
It seems we only have something like this in EQ.

Reply:
Yeah, I was thinking we could make it generic to combine this and handle the hazard layer too.